### PR TITLE
chore: move source data transformation from data forwarder to sourcer

### PR DESCRIPTION
**Pull request by KeranYang**
This change moves source data transformer from the data forwarder to the sourcer. By doing so we make sure we assign source watermarks according to the updated message event times.

The following related changes are included:

* Add de-duplication logics to data forwarder to ensure the offset list we sent to BufferReader Ack() doesn't have duplicates.

* Add PublishSourceWatermarks method to sourcer interface to ensure future sourcers all take care of publishing watermarks.

* Add a simple e2e test to verify watermark propagation after source data transformation.

Signed-off-by: Keran Yang <yangkr920208@gmail.com>

Explain what this PR does.

### CODE REVIEW

File: pkg/forward/forward.go [hunk 1]
```diff
@@ -334,7 +334,10 @@ func (isdf *InterStepDataForward) ackFromBuffer(ctx context.Context, offsets []i
 		Steps:    math.MaxInt,
 		Duration: time.Millisecond * 10,
 	}
-	var ackOffsets = offsets
+
+	// we only need to acknowledge every distinct offset once, for now, we add de-duplication logics here before sending offsets to fromBuffer.Ack().
+	// as of now, we don't force fromBuffer to implement Ack() as an idempotent function, but we should for future implementations.
+	var ackOffsets = isb.DeduplicateOffsets(offsets)
```
Comment (KeranYang) [inline pkg/forward/forward.go:340]: This introduces extra O(n) time (and space) complexity for every batch. The time could be saved by moving the dedup logic to outside of ackFromBuffer.

Simply change the following code 

```
var readOffsets = make([]isb.Offset, len(readMessages))
for idx, m := range udfResults {
	readOffsets[idx] = m.readMessage.ReadOffset
}
err = isdf.ackFromBuffer(ctx, readOffsets)
```

to

```
set := make(map[string]bool)
var readOffsets []isb.Offset
for _, m := range udfResults {
	if ok := set[m.readMessage.ReadOffset.String()]; !ok {
		set[m.readMessage.ReadOffset.String()] = true
		readOffsets = append(readOffsets, m.readMessage.ReadOffset)
	}
}
```

But the O(n) space complexity for map still exist.

Comment (vigith) [reply pkg/forward/forward.go:340]: jfyi, you can replace `bool` with `struct{}` that takes 0 space. we are just looking for the presence/absence of a key so the value shouldn't matter.

**Comment by KeranYang** [comment]
@vigith @whynowy , Derek had a new proposal on this task. I played around with it and I was convinced that it could be the better approach. See below for comparison between current and proposed approach. The implementation of the new approcah is also in PR, please review. Let's work together to make a choice.

<!DOCTYPE html><h3 cid="n0" mdtype="heading" class="md-end-block md-heading" style="box-sizing: border-box; white-space: pre-wrap; break-after: avoid-page; break-inside: avoid; orphans: 4; font-size: 1.5em; margin-top: 1rem; margin-bottom: 1rem; position: relative; font-weight: bold; line-height: 1.43; cursor: text; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-style: normal; font-variant-caps: normal; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="plain" class="md-plain md-expand" style="box-sizing: border-box;">Requirements</span></h3><ol class="ol-list" cid="n4" mdtype="list" style="box-sizing: border-box; margin: 0.8em 0px; padding-left: 30px; position: relative; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><li class="md-list-item" cid="n6" mdtype="list_item" style="box-sizing: border-box; margin: 0px; position: relative;"><p cid="n7" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0px 0px 0.5rem; white-space: pre-wrap; position: relative;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Message IsLate attribute gets assigned </span><span md-inline="strong" class="md-pair-s " style="box-sizing: border-box;"><strong style="box-sizing: border-box;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">AFTER</span></strong></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> source data transformation.</span></p></li><li class="md-list-item" cid="n10" mdtype="list_item" style="box-sizing: border-box; margin: 0px; position: relative;"><p cid="n8" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0px 0px 0.5rem; white-space: pre-wrap; position: relative;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Source Watermark gets published </span><span md-inline="strong" class="md-pair-s " style="box-sizing: border-box;"><strong style="box-sizing: border-box;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">AFTER</span></strong></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> source data transformation.</span></p></li><li class="md-list-item md-focus-container" cid="n56" mdtype="list_item" style="box-sizing: border-box; margin: 0px; position: relative;"><p cid="n54" mdtype="paragraph" class="md-end-block md-p md-focus" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0px 0px 0.5rem; white-space: pre-wrap; position: relative;"><span md-inline="plain" class="md-plain md-expand" style="box-sizing: border-box;">IsLate attribute get assigned </span><span md-inline="strong" class="md-pair-s " style="box-sizing: border-box;"><strong style="box-sizing: border-box;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">BEFORE</span></strong></span><span md-inline="plain" class="md-plain" style="box-sizing: border-box;"> writing to toBuffer ISBs.</span></p></li></ol><h3 cid="n2" mdtype="heading" class="md-end-block md-heading" style="box-sizing: border-box; white-space: pre-wrap; break-after: avoid-page; break-inside: avoid; orphans: 4; font-size: 1.5em; margin-top: 1rem; margin-bottom: 1rem; position: relative; font-weight: bold; line-height: 1.43; cursor: text; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-style: normal; font-variant-caps: normal; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Current Approach</span></h3><p cid="n12" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0.8em 0px; white-space: pre-wrap; position: relative; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Sourcer reads 500 messages, applies transformer, publishes source watermarks and returns to forwarder the transformed messages (could be more than 500 messages). Forwarder assigns IsLate, forwards the messages, de-duplicates them by ReadOffset and acknowledges.</span></p><p cid="n20" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0.8em 0px; white-space: pre-wrap; position: relative; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="url" class="md-link md-pair-s" spellcheck="false" style="box-sizing: border-box; word-break: break-all;"><a href="https://github.com/numaproj/numaflow/pull/544" style="box-sizing: border-box; cursor: pointer; color: rgb(65, 131, 196); -webkit-user-drag: none;">https://github.com/numaproj/numaflow/pull/544</a></span></p><h3 cid="n14" mdtype="heading" class="md-end-block md-heading" style="box-sizing: border-box; white-space: pre-wrap; break-after: avoid-page; break-inside: avoid; orphans: 4; font-size: 1.5em; margin-top: 1rem; margin-bottom: 1rem; position: relative; font-weight: bold; line-height: 1.43; cursor: text; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-style: normal; font-variant-caps: normal; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Proposed Approach</span></h3><p cid="n16" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0.8em 0px; white-space: pre-wrap; position: relative; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Sourcer reads 500 messages and passes them to forwarder (exactly 500 messages). Forwarder uses existing UDF implementation to apply transformer. After transformation, fowarder assigns IsLate, publishes source watermarks (using source watermark publisher provided by New()), forwards the messages and acknowledges.</span></p><p cid="n64" mdtype="paragraph" class="md-end-block md-p" style="box-sizing: border-box; line-height: inherit; orphans: 4; margin: 0.8em 0px; white-space: pre-wrap; position: relative; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="url" class="md-link md-pair-s" spellcheck="false" style="box-sizing: border-box; word-break: break-all;"><a href="https://github.com/numaproj/numaflow/pull/557" style="box-sizing: border-box; cursor: pointer; color: rgb(65, 131, 196); -webkit-user-drag: none;">https://github.com/numaproj/numaflow/pull/557</a></span></p><h3 cid="n61" mdtype="heading" class="md-end-block md-heading" style="box-sizing: border-box; white-space: pre-wrap; break-after: avoid-page; break-inside: avoid; orphans: 4; font-size: 1.5em; margin-top: 1rem; margin-bottom: 1rem; position: relative; font-weight: bold; line-height: 1.43; cursor: text; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-style: normal; font-variant-caps: normal; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;"><span md-inline="plain" class="md-plain" style="box-sizing: border-box;">Why proposed approach is better (in my opinion)</span></h3><figure class="md-table-fig" cid="n39" mdtype="table" style="box-sizing: border-box; margin: 1.2em 0px; overflow-x: auto; max-width: calc(100% + 16px); padding: 0px; cursor: default; caret-color: rgb(51, 51, 51); color: rgb(51, 51, 51); font-family: &quot;Open Sans&quot;, &quot;Clear Sans&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;Segoe UI Emoji&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration: none;">

Pros. | Cons.
-- | --
Re-use the existing UDF. No need to implement the entire gRPC call for transformer. Also save future efforts if we want to make updates to UDF implemenations as we don't need to replicate the change to transformer. (Simpler) |  
No need to de-duplicate ReadOffsets for acknowledging. (Simpler) |  
Source watermark publish logic gets moved out of the BufferReader Read() method, which follows single responsibility principle. (Clearer) | Move the source watermark publish execution to forwarder. Non-source forwarders also need to specify a nil as source watermark publisher.
  |  

</figure>

**Comment by vigith** [comment]
> Move the source watermark publish execution to forwarder. Non-source forwarders also need to specify a nil as source watermark publisher.

Can't we do a `IsSource()` check in the forwarder so that non-source forwarders do not have to specify a nil?

**Comment by KeranYang** [comment]
> Can't we do a `IsSource()` check in the forwarder so that non-source forwarders do not have to specify a nil?

Yes, that's a good idea. I implemented something similar in the PR where I made srcWatermarkPublisher as optional by putting it to options.

**Comment by KeranYang** [comment]
Closing this PR as we decided to go with the new proposed approach. #557

**Pull request closed without merging**